Added WebSocket connection documentation (#3393)
hoprnet/hoprnet: docs/hopr-documentation/docs/developers/connecting-node.md

---
id: connecting-node
title: Interacting with a HOPR node
---

Once you have your HOPR cluster up and running, you are ready to connect to a single HOPR node and listen to messages sent to it by other HOPR nodes.
The following guide will show you how to connect to a running HOPR node, and verify some basic functionality using both the REST and WebSocket
endpoints.

## Connect to a HOPR node

### 1. Get the node security credentials

To avoid unsecured access to your HOPR node, all HOPR node's WebSocket and REST calls require an `apiToken`. Your API token needs to be appended as a
`query` parameter for WebSocket connections and as encoded token via the `Authorization` header for you to be able to connect to it.

:::info Tip
The flag used to set this value in a HOPR node via `hoprd` is `--apiToken`. The default `apiToken` used across our documentation is the following one 
(click on the right to copy), so make sure to change it when running your own node in a public network.
<br/>

```
^^LOCAL-testing-123^^
```
:::

Export your `apiToken` to be used in the incoming commands.

```
export apiToken="^^LOCAL-testing-123^^"
```

### 2. Test REST API connectivity

> Given your node's `apiToken` and your `HOPR_NODE_1_HTTP_URL` from the last section, try to send a REST command to query it’s address with the following `curl`

Reviewer: Minor thing: I would include a line saying e.g. `HOPR_NODE1_HTTP_URL=localhost:3001` just for clarity and reminder. Also because of the port number that needs to be included in the URL and might not be immediately obvious.

Author: ```suggestion
Given your node's `apiToken` and your `HOPR_NODE_1_HTTP_URL` from the last section (likely `localhost:3001`), try to send a REST command to query its address with the following `curl`
```

Good stuff, let me use the suggestion feature and do the change directly.

command. If you don’t have `jq` installed, just remove it at the end of the command.

```
echo -n $apiToken | base64 | xargs -I {} curl -s -H "Authorization: Basic {}" $HOPR_NODE_1_HTTP_URL/api/v2/account/address | jq
```

If successful, you should get a response similar to this one:

```
{
  "nativeAddress": "0x3a54dDE3ee5ACfd43C902cbecC8ED0CBA10Ff326",
  "hoprAddress": "16Uiu2HAmE9b3TSHeF25uJS1Ecf2Js3TutnaSnipdV9otEpxbRN8Q"
}
```

In case you have made a mistake, like forgotten to use `-n` in your `echo` or have the wrong `apiToken`, you’ll see the following instead:

```
{
  "status": 403,
  "challenge": "Basic realm=hoprd",
  "message": "You must authenticate to access hoprd."
}
```

### 3. Test WebSocket connectivity

Unlike our REST API endpoint, to see interactions against our HOPR node WebSocket server, we need to use a client to both send and listen
connections in our HOPR node. Our suggested client is [websocat](https://github.com/vi/websocat), which you can install by running our

> `./scripts/install-websocat.sh` [script](https://raw.githubusercontent.com/hoprnet/hoprnet/master/scripts/install-websocat.sh) from our monorepo.

Reviewer: Maybe a note that this script installs `websocat` to `../.bin` directory relative to `./scripts`.

Author: Yeah, that's really annoying, but noted.

Author: ```suggestion
`./install-websocat.sh` [script](https://raw.githubusercontent.com/hoprnet/hoprnet/master/scripts/install-websocat.sh) from our monorepo.
To install, make sure to run it from the `scripts` folder, as it will install it in the parent folder under a `.bin` folder, and will not export it to your `$PATH`.
```

:::warning @TODO
Migrate `install-websocat.sh` script to allow one-line setup via bash pipeing
```

> curl -o- https://raw.githubusercontent.com/hoprnet/hoprnet/master/scripts/install-websocat.sh | bash

Reviewer: Also beware of where the script installs the binary, see above comment.

Author: Yeah, this is more of todo for us later on, we need to refactor this script so it,

- can be installed by being pipped out to our bash (right now it fails due to `log`)
- doesn't expect to be run within the `scripts` folder, neither install it in our `.bin` folder
- exposes it to the local `$PATH` so it shows after `which websocat`

Creating an issue for this in particular.

Author: Documented in https://github.com/hoprnet/hoprnet/issues/3405

<details>
  <summary>Export REST and WebSocket endpoints (from local HOPR cluster)</summary>
  <div>
    <div>
    <h3>Node 1</h3>
    <br/>
    <pre>
    export HOPR_NODE_1_HTTP_URL=http://127.0.0.1:13301 HOPR_NODE_1_WS_URL=http://127.0.0.1:19501
    </pre>
    <h3>API token & Node 1</h3>
    <br/>
    <pre>
    export apiToken=^^LOCAL-testing-123^^ HOPR_NODE_1_HTTP_URL=http://127.0.0.1:13301 HOPR_NODE_1_WS_URL=http://127.0.0.1:19501
    </pre>
    </div>
    <br/>
  </div>
</details>

<details>
  <summary>Export REST and WebSocket endpoints (from single localhost HOPR node)</summary>
  <div>
    <div>
    <h3>Node 1</h3>
    <br/>
    <pre>
    export HOPR_NODE_1_HTTP_URL=http://127.0.0.1:3001 HOPR_NODE_1_WS_URL=http://127.0.0.1:3000
    </pre>
    <h3>API token & Node 1</h3>
    <br/>
    <pre>
    export apiToken=^^LOCAL-testing-123^^ HOPR_NODE_1_HTTP_URL=http://127.0.0.1:3001 HOPR_NODE_1_WS_URL=http://127.0.0.1:3000
    </pre>
    </div>
    <br/>
  </div>
</details>

We use `127.0.0.1` instead of `localhost` as some tools like `websocat` (described below) struggle to resolve `localhost` properly.

### 3. Install a WebSocket and REST client

> To properly[^1] interact with a HOPR node, you need both a REST and a WebSocket client software. A REST client is used to give instructionst to our

Reviewer: Minor thing: we do I need BOTH WS and REST client? If I can just send commads via REST and get my success/error responses back I can be happy, right?

Author: You do need for receiving messages on the node you are sending messages TO. So yes, if you are interacting with `node 1`, you only need the REST client, but for reading the message `node 1` sends, (e.g. sends a message to `node 2`), you need a WS client connected to `node 2` to see it. I made that comment in the footnote at the end, connected to the word `properly`.
